# Boston Housing Assignment

Implimenting scikit learn's r2 score and mse methods for measuring the performance of the linear regressor and sklearn.linear_model.Ridge modle

In [1]:
from sklearn import datasets
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression

In [2]:
bean = datasets.load_boston()
print bean.DESCR

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

In [3]:
def load_boston():
    scaler = StandardScaler()
    boston = datasets.load_boston()
    X=boston.data
    y=boston.target
    X = scaler.fit_transform(X)
    return train_test_split(X,y)

In [4]:
X_train, X_test, y_train, y_test = load_boston()

In [5]:
X_train.shape

(379L, 13L)

## Fitting a Linear Regression

This is done by instantiating a new regression object (line 1) and giving your regression object your training data (line 2) by calling .fit(independent variables, dependent variable)

In [6]:
clf = LinearRegression()
clf.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

## Making a Prediction

In [7]:
zip (y_test, clf.predict(X_test))

[(50.0, 24.890270701428285),
 (10.800000000000001, 11.90937468923655),
 (11.9, 8.7310837739480274),
 (24.399999999999999, 28.224110941659468),
 (18.199999999999999, 18.428802983159272),
 (10.4, 17.07615288403619),
 (16.300000000000001, 12.550384531771781),
 (33.299999999999997, 35.24421658907184),
 (23.0, 23.656695451319511),
 (35.399999999999999, 30.417651817349956),
 (38.700000000000003, 35.741872239061081),
 (16.100000000000001, 18.692775035166672),
 (19.800000000000001, 22.555986095105375),
 (19.399999999999999, 19.82119911395387),
 (22.199999999999999, 22.290343309308135),
 (21.699999999999999, 25.098777757847728),
 (11.699999999999999, 16.036538801705625),
 (17.199999999999999, 17.228690466264709),
 (37.200000000000003, 32.976893021417119),
 (27.899999999999999, 31.070255421803644),
 (24.600000000000001, 24.896503396587104),
 (33.100000000000001, 35.475483698668342),
 (35.100000000000001, 35.216181125440045),
 (20.699999999999999, 26.618040187808052),
 (13.300000000000001, 20.619

In [8]:
mean_squared_error(y_test,clf.predict(X_test))

24.335492657229299

In [9]:
r2_score(y_test,clf.predict(X_test))

0.74715765225195818

## MSE = 24.335492657229299

## r2_score = 0.74715765225195818

## Computing Linear Regression

taking alpha=0.001 for computing sklearn.linear_model.Ridge


In [10]:
from sklearn.linear_model import Ridge
r = Ridge(alpha=.001)
r.fit(X_train, y_train)

Ridge(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [11]:
zip (y_test, r.predict(X_test))

[(50.0, 24.890209902574355),
 (10.800000000000001, 11.909388005786369),
 (11.9, 8.7310457142463047),
 (24.399999999999999, 28.224061832538787),
 (18.199999999999999, 18.428822292239978),
 (10.4, 17.076257381372731),
 (16.300000000000001, 12.550355480825024),
 (33.299999999999997, 35.2442241038422),
 (23.0, 23.656643325672515),
 (35.399999999999999, 30.417623714051022),
 (38.700000000000003, 35.741886435591717),
 (16.100000000000001, 18.692769644243462),
 (19.800000000000001, 22.556008282157613),
 (19.399999999999999, 19.821226809183521),
 (22.199999999999999, 22.290358973718639),
 (21.699999999999999, 25.098749137681626),
 (11.699999999999999, 16.036542124605635),
 (17.199999999999999, 17.228661768424736),
 (37.200000000000003, 32.97684910193675),
 (27.899999999999999, 31.070179826149598),
 (24.600000000000001, 24.896518335374392),
 (33.100000000000001, 35.4754425620036),
 (35.100000000000001, 35.216103752672851),
 (20.699999999999999, 26.617999593347335),
 (13.300000000000001, 20.6193

In [12]:
mean_squared_error(y_test,r.predict(X_test))

24.335503991903398

In [13]:
r2_score(y_test,r.predict(X_test))

0.74715753448628763

## MSR = 24.335503991903398

## r2_score = 0.74715753448628763


# Ridge Linear Regression



### taking alpha =2

In [14]:
from sklearn.linear_model import Ridge
r = Ridge(alpha=2)
r.fit(X_train, y_train)

Ridge(alpha=2, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [15]:
zip (y_test, r.predict(X_test))

[(50.0, 24.772887160038643),
 (10.800000000000001, 11.936252221499981),
 (11.9, 8.6593171462388359),
 (24.399999999999999, 28.131338973038595),
 (18.199999999999999, 18.46580862267318),
 (10.4, 17.268983263669746),
 (16.300000000000001, 12.494560051867639),
 (33.299999999999997, 35.25603424992012),
 (23.0, 23.557214467467944),
 (35.399999999999999, 30.367684695112217),
 (38.700000000000003, 35.764408679184328),
 (16.100000000000001, 18.683435173108901),
 (19.800000000000001, 22.595310464152526),
 (19.399999999999999, 19.871583730511038),
 (22.199999999999999, 22.323382008600277),
 (21.699999999999999, 25.044557133877056),
 (11.699999999999999, 16.043520846321911),
 (17.199999999999999, 17.173958626292738),
 (37.200000000000003, 32.891810828961283),
 (27.899999999999999, 30.927124554407921),
 (24.600000000000001, 24.925415446506761),
 (33.100000000000001, 35.396228547226521),
 (35.100000000000001, 35.06858373102375),
 (20.699999999999999, 26.541517045791146),
 (13.300000000000001, 20.58

In [16]:
mean_squared_error(y_test,r.predict(X_test))

24.361937550142731

In [17]:
r2_score(y_test,r.predict(X_test))

0.74688289353207704

## MSR = 24.361937550142731

## r2_score = 0.74688289353207704

Thus we see that when value of alpha is decreased to 0.001 there is not much difference in MSR and R2 values.
When we change the value of alpha to 2 the value of MSR changes very slightly.However R2 chanes a bit